osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/threads/RecoverKeyProcess.java

```diff
@@ -129,65 +100,28 @@ private void checkState() {
     }
   }
 
-  private boolean canConnectUsingNewKeys() {
+  private boolean canConnectUsingNewKeys(final DlmsDevice device) {
     DlmsConnection connection = null;
     try {
-      connection = this.createConnectionUsingNewKeys();
-      return true;
+      this.throttlingService.openConnection();
+
+      connection =
+          this.hls5Connector.connectUnchecked(
+              device, null, this.secretManagementService::getNewKeys);
+      return connection != null;
     } catch (final Exception e) {
-      LOGGER.warn("Connection exception: {}", e.getMessage(), e);
+      log.error("Connection exception: {}", e.getMessage(), e);
       return false;
     } finally {
+      this.throttlingService.closeConnection();
+
       if (connection != null) {
         try {
           connection.close();
         } catch (final IOException e) {
-          LOGGER.warn("Connection exception: {}", e.getMessage(), e);
+          log.warn("Connection exception: {}", e.getMessage(), e);
         }
       }
     }
   }
```

Comment:
Something about this code still worries me.

The thottling service is used to maintain the guards that it provides, but the set up falls short on other aspects that play a roll.

Compare this code for instance to what happens in `org.opensmartgridplatform.adapter.protocol.dlms.infra.messaging.DeviceRequestMessageProcessor` from `processMessage(ObjectMessage)`.

As part of the set up for the connection this calls `DlmsConnectionMessageProcessor` where from `createConnectionForDevice` a `DlmsMessageListener` is instanitiated. In the use case with the `RecoverKeyProcess` there is a good chance that this would be an instance of the `InvocationCountingDlmsMessageListener`.

Back in `DeviceRequestMessageProcessor` after the connection was used there is `doConnectionPostProcessing`  where the invocation counter is updated.

Even just opening the connection as happens here in `canConnectUsingNewKeys` causes the invocation counter to be incremented a little (adding 2 to the previous value).

If we should make an effort not to get out of sync, the code for the connection here may need to be extended to also include properly maintaining the invocation counter value for the device.

Reply:
I've added code to keep the invocation counter in sync